[ews-build.webkit.org] Handle duplicate GitHub hooks

JonWBedard opened the pull request:
#### a2343bf6a44c76e9fff85a24d9fb69a559c6cfb4
<pre>
[ews-build.webkit.org] Handle duplicate GitHub hooks
<a href="https://bugs.webkit.org/show_bug.cgi?id=269034">https://bugs.webkit.org/show_bug.cgi?id=269034</a>
<a href="https://rdar.apple.com/122589894">rdar://122589894</a>

Reviewed by Aakash Jain.

Create a process-local record of events from GitHub, and prevent
events duplicated within 60-120 seconds of each over from
triggering duplicate builds.

* Tools/CISupport/ews-build/events.py:
(GitHubEventHandlerNoEdits.is_duped): Check if a value was registered
in the last two 60 second buckets.
(GitHubEventHandlerNoEdits.handle_pull_request): Before dispatching builds,
ensure that the incoming event hasn&apos;t been duplicated in the few minutes.
* Tools/CISupport/ews-build/events_unittest.py:
(TestCommitClassifier.test_duplicate):

Canonical link: <a href="https://commits.webkit.org/274369@main">https://commits.webkit.org/274369@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/252b97140d7a7c77f229f48f9f84da8d77798aa9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38906 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17837 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41251 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41437 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41213 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20690 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15186 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39479 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15006 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33733 "Build was cancelled. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (cancelled)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13042 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/34655 "Build was cancelled. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 7 flakes 1 failures; Uploaded test results; layout-tests (exception)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42714 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35303 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38825 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13716 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11311 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37055 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/38954 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/15323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/14986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5072 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -356,7 +356,25 @@ class GitHubEventHandlerNoEdits(GitHubEventHandler):
 
+    # This approach only works as long as we have a single build-master responding to hooks
+    @classmethod
+    def is_duped(cls, value, bucket=60):
```
Might be a good idea to add some comments in this function describing what/how we are doing here.

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -512,20 +530,26 @@ def handle_pull_request(self, payload, event):
+
+        log.msg(f'Handling PR #{pr_number} with hash: {head_sha}')
+        if event in ('unsafe_merge_queue', 'merge_queue'):
+            yield task.deferLater(reactor, self.LABEL_PROCESS_DELAY, lambda: None)
```
what exactly does this line do?

JonWBedard replied:
This line is important because the first thing we do when processing a `merge-queue` build is check if `merge-queue` is still on the PR. In practice, we've noticed that GitHub's labels take time to propagate in GitHub's infrastructure. Put another way, it's possible to receive an event that a label has been added, but then go check the PR and have GitHub tell you the PR does _not_ have the label yet.

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -512,20 +530,26 @@ def handle_pull_request(self, payload, event):
+        # We received this exact event in the last 60 seconds, ignore it
+        if cls.is_duped(f'{event}-{head_sha}'):
+            log.msg(f'Discarding duplicate for PR #{pr_number} with hash: {head_sha}')
+            return defer.returnValue(([], 'git'))
```
Independent from this change, it would be helpful to add logging indicating that we are going to process this change. For case when we ignore the PR we/buildbot logs message about ignoring, but would be good to have a message about success case as well. Maybe a log just before returning from this method.

JonWBedard replied:
We have that! The next statement is:
```
log.msg(f'Handling PR #{pr_number} with hash: {head_sha}')
```
the bit we aren't logging is the classification of the commits in a PR. Which I think is fine, that's visible in the build steps, and it's really the build steps that need this anyways. The only reason we compute classification upon hook reception is that it greatly reduces the number of times we have to compute classification.

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -512,20 +530,26 @@ def handle_pull_request(self, payload, event):
             event = 'merge_queue'
 
+        # We received this exact event in the last 60 seconds, ignore it
+        if cls.is_duped(f'{event}-{head_sha}'):
```
This doesn't really seem thread-safe. If two duplicate events from GitHub come at same time, is_duped() can still return False for both of them and continue to duplicate builds.

We can either make it thread-safe. Or for urgent fix one work-around can be to move this check little later, after `yield super().handle_pull_request()`. Since that method in base class perform bunch of network requests, it would most likely add a different amount of delays for the two requests which came at same time.

JonWBedard replied:
It's not thread-safe, but it doesn't need to be.

Twisted only has a single thread that's it's doing all it's work in, so as long as we only have a single web server, we can guarantee that PRs will be processed one at a time. Note that I deployed this on the staging instance and we saw exactly that happen.

JonWBedard committed:
[ews-build.webkit.org] Handle duplicate GitHub hooks
https://bugs.webkit.org/show_bug.cgi?id=269034
rdar://122589894

Reviewed by Aakash Jain.

Create a process-local record of events from GitHub, and prevent
events duplicated within 60-120 seconds of each over from
triggering duplicate builds.

* Tools/CISupport/ews-build/events.py:
(GitHubEventHandlerNoEdits.is_duped): Check if a value was registered
in the last two 60 second buckets.
(GitHubEventHandlerNoEdits.handle_pull_request): Before dispatching builds,
ensure that the incoming event hasn't been duplicated in the few minutes.
* Tools/CISupport/ews-build/events_unittest.py:
(TestCommitClassifier.test_duplicate):

Canonical link: https://commits.webkit.org/274369@main
